### ISSUE TITLE
Support custom venv locations for `uv sync`

### ISSUE THREAD

**Issue by andriygm** [astral-sh/uv#5229]
`uv pip install` supports the `VIRTUAL_ENV` environment variable, allowing package installation into venvs not located in the same directory as the project. `uv sync` would also benefit from being able to install packages into external venvs (perhaps as a flag?)

Since `uv sync` also creates venvs if they don't exist, it should do so in the custom location.

**Comment by zanieb** [comment]
Related https://github.com/astral-sh/rye/issues/1211

**Comment by nazq** [comment]
> Related [astral-sh/rye#1211](https://github.com/astral-sh/rye/issues/1211)

÷1 and thanks for linking the related rye enhancement

**Comment by zanieb** [comment]
@charliermarsh I don't think this should be in scope for the first round of stabilizations.

**Comment by tiangolo** [comment]
Here's use case I have where it would be useful. I was trying to migrate https://github.com/fastapi/full-stack-fastapi-template/ to `uv`.

I mount the local development directory inside the Docker container as a volume: https://github.com/fastapi/full-stack-fastapi-template/blob/master/docker-compose.override.yml#L58

This allows fast iteration on the code, as it only takes a reload of the server instead of a re-build of the Docker image to try a change.

If I have a local `.venv` directory, it will end up mounted as well. But inside the container, the file links are not resolved.

I tried using `UV_SYSTEM_PYTHON` and `VIRTUAL_ENV` to move it to some root directory (so that it wouldn't be mounted), but `uv` would still try to use the venv in `.venv`. And if I remove it from inside the container, the in-container `uv` would create a new venv in `.venv` that would also show up in the local (mounted directory).

---

Somewhat related, I would like to be able to use Python directly, not only through `uv run`. For example, in that project, I currently use a base Docker image that comes with its own entrypoint and command. But those wouldn't be able to run the programs without activating the `uv` venv in some way (or using the system Python).

**Comment by charliermarsh** [comment]
Makes a ton of sense, thanks @tiangolo. I think this is pretty high-priority now that the release is out.

> Somewhat related, I would like to be able to use Python directly, not only through uv run.

Here, you're referring to the installed Python, like outside of the virtualenv, is that right? (Since you can always `source .venv/bin/activate` without going through `uv run`.)

**Comment by tiangolo** [comment]
> Makes a ton of sense, thanks @tiangolo. I think this is pretty high-priority now that the release is out.

:rocket: :tada: 

> Here, you're referring to the installed Python, like outside of the virtualenv, is that right? (Since you can always source .venv/bin/activate without going through uv run.)

Yep, this part was a misunderstanding and mistake from my side. :sweat_smile:

I was misinterpreting the docs. And then I was trying something that wouldn't work.

But this worked:

```Dockerfile
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10
# Install uv
COPY --from=ghcr.io/astral-sh/uv:0.3.0 /uv /bin/uv

WORKDIR /app/

ENV VIRTUAL_ENV=/app/.venv
# Place executables in the environment at the front of the path
ENV PATH="/app/.venv/bin:$PATH"

# Maybe copy lock file in case it doesn't exist in the repo
COPY ./pyproject.toml ./uv.lock* /app/

RUN uv sync

RUN . $VIRTUAL_ENV/bin/activate
```

**Note**: it seems the build steps with `RUN` are run through `sh`, not `bash`, so the `source` command is not available, only the `.` syntax (`. .venv/bin/activate`).

<details>
<summary>full Dockerfile</summary>

```Dockerfile
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10
# Install uv
COPY --from=ghcr.io/astral-sh/uv:0.3.0 /uv /bin/uv

WORKDIR /app/

ENV VIRTUAL_ENV=/app/.venv
# Place executables in the environment at the front of the path
ENV PATH="/app/.venv/bin:$PATH"

# Maybe copy lock file in case it doesn't exist in the repo
COPY ./pyproject.toml ./uv.lock* /app/

# Allow installing dev dependencies to run tests
ARG INSTALL_DEV=false

RUN bash -c "if [ $INSTALL_DEV == 'true' ] ; then uv sync ; else uv sync --no-dev ; fi"

RUN . $VIRTUAL_ENV/bin/activate

ENV PYTHONPATH=/app

COPY ./scripts/ /app/

COPY ./alembic.ini /app/

COPY ./prestart.sh /app/

COPY ./tests-start.sh /app/

COPY ./app /app/app
```

</details>

**Comment by lukewiwa** [comment]
To add to this. Being able to configure the `.venv` location will also help out when using docker bind mounts during the build phase. https://docs.docker.com/build/building/best-practices/#add-or-copy. Currently this is impossible in `uv` since the bind mount is read only and `uv` will try and write to the mount and error out due to attempting to write to the mount.

eg 
```dockerfile
FROM python:3.12

COPY --from=ghcr.io/astral-sh/uv:0.3.0 /uv /usr/local/bin/uv

# Assuming a python project in the directory `project_name` with a `pyproject.toml` file
ENV UV_SYSTEM_PYTHON=true VIRTUAL_ENV=/opt/.venv
RUN --mount=type=bind,source=project_name,target=/project_name \
  uv sync
```

Which will result in the following error
```
error: failed to create directory `/project_name/.venv`
Caused by: Read-only file system (os error 30)
```

**Comment by zanieb** [comment]
@tiangolo just a heads up that providing a `python` shim is tracked in https://github.com/astral-sh/uv/issues/6265

Note we also do cover some of this in our [Docker integration guide](https://docs.astral.sh/uv/guides/integration/docker/) but clearly there's room for improvement in the Docker integration story.

**Comment by tiangolo** [comment]
Thanks @zanieb! I think what I needed would be covered in some way by activating the environment. I would like to be able to `UV_SYSTEM_PYTHON=true`, but at least putting the virtual environment in another location would work.

---

I personally actually don't like shims, I can't use `which python` with shims, that's the main reason I ended up not using other tools that used/needed shims. :sweat_smile: 

---

About Docker, yep! Actually great docs! That's what I was basing my work on. Maybe the detail of `RUN . $VIRTUAL_ENV/bin/activate` could be added, don't know. :thinking: 

But anyway, great job on the docs already. :clap:

**Comment by sbidoul** [comment]
Related, re sync in dockerfile: https://github.com/astral-sh/uv/issues/4028#issuecomment-2288187425

**Comment by adiberk** [comment]
> > Makes a ton of sense, thanks @tiangolo. I think this is pretty high-priority now that the release is out.
> 
> 🚀 🎉
> 
> > Here, you're referring to the installed Python, like outside of the virtualenv, is that right? (Since you can always source .venv/bin/activate without going through uv run.)
> 
> Yep, this part was a misunderstanding and mistake from my side. 😅
> 
> I was misinterpreting the docs. And then I was trying something that wouldn't work.
> 
> But this worked:
> 
> ```dockerfile
> FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10
> # Install uv
> COPY --from=ghcr.io/astral-sh/uv:0.3.0 /uv /bin/uv
> 
> WORKDIR /app/
> 
> ENV VIRTUAL_ENV=/app/.venv
> # Place executables in the environment at the front of the path
> ENV PATH="/app/.venv/bin:$PATH"
> 
> # Maybe copy lock file in case it doesn't exist in the repo
> COPY ./pyproject.toml ./uv.lock* /app/
> 
> RUN uv sync
> 
> RUN . $VIRTUAL_ENV/bin/activate
> ```
> 
> **Note**: it seems the build steps with `RUN` are run through `sh`, not `bash`, so the `source` command is not available, only the `.` syntax (`. .venv/bin/activate`).
> 
> full Dockerfile

@tiangolo am actually having a similar issue except we install to the system python in the Dockerfile -  so I have maade a similar request to allow disabling force install into virtual env..  [Here is link](https://github.com/astral-sh/uv/issues/6459)
However doing RUN . $VIRTUAL_ENV/bin/activate doesn't seem to help in my case

Happy it is working for you!

Also love the project team!

**Comment by zanieb** [comment]
@adiberk — yeah that won't work. See our [documentation](https://docs.astral.sh/uv/guides/integration/docker/#installing-a-project) for a recommendation on how to activate the environment properly.

**Comment by adiberk** [comment]
I figured it out
If I do everything inside a different workdir like /opt/
And activate the env from there as well as set the python path to there, it works exactly as it should.

I assume my issue is that the virtual env is getting messed up by the way we mount our volume as well as copy our folders in the Dockerfile (one or the other)

**Comment by zanieb** [comment]
Does adding the `.venv` to your project's `.dockerignore` fix that too?

**Comment by talhajunaidd** [comment]
Here is my solution for docker file

```
ENV UV_NO_CACHE=1
ENV UV_SYSTEM_PYTHON=true
ENV UV_PYTHON=/usr/local/bin/python
RUN uv pip compile pyproject.toml -o requirements.txt && uv pip sync --system requirements.txt
```

**Comment by adiberk** [comment]
> Does adding the `.venv` to your project's `.dockerignore` fix that too?

It didn’t in my case.

**Comment by hauntsaninja** [comment]
Similar to https://github.com/astral-sh/uv/issues/1495#issuecomment-1950442191 , you can symlink to solve this.

```
uv venv /app/.venv
ln -sf /app/.venv .venv
uv sync
```

**Comment by jklaiho** [comment]
Just to chime in with our use case: currently, we use pip-tools in our container images, where we have `VIRTUAL_ENV=/opt/venv` and `PATH=/opt/venv/bin:$PATH`. We have an intricate multi-stage Dockerfile that populates `/opt/venv` in an early build stage using `pip-sync` on a compiled requirement file, and it's `COPY`'d wholesale to the actual runtime stage with no build stage dependencies present. Would be nice to run `uv sync` instead of `pip-sync` in the build stage to achieve the same, migrate from a compiled requirement file to `uv.lock`, and change nothing else about how our system currently works.

**Comment by zanieb** [comment]
> > Does adding the `.venv` to your project's `.dockerignore` fix that too?
> 
> It didn’t in my case.

Can you elaborate here? It still mounted `.venv` into the container?

**Comment by adiberk** [comment]
I can retest using your suggestion, however I recall in my original tests (before these comments) I had actually done that (with dockerignore) and it didn’t help. Post these comments, I simply tested by removing the .venv folder from my local and tested the docker setup suggested and still had issues. 
The only thing that worked for me was running uv sync with my lockfile in a different workdir. And then setting env in my path.

**Comment by edmorley** [comment]
Another use-case for having a custom venv location outside of the project directory is CNBs ([Cloud Native Buildpacks](https://buildpacks.io/)).

With CNBs, the layer/caching model is different to `Dockerfile`, so standard Dockerfile best practices don't apply. For example, with CNBs:
- buildpacks aren't run as root
- layers are created as directories under `/layers/<buildpack_name>/<layer_name>` which are separate to each other, and also separate to the app source code
- layers can be marked with various properties that control whether they are cached or made available in the final run-image or not

These properties allow for several benefits such as base image "rebase", finer grained cache re-use/invalidation, smaller run-images (via excluding build time only deps), the ability to support multiple languages/ecosystems (Docker multi-stage builds get messy fast when trying to build a multi-language app, since it's hard to re-use official Docker images etc).

However, they do mean that the various components have to be installed into separate directory trees (rather than relying on an overlay filesystem to keep files contributed by different layers separate).

For example when using pip a possible layout would be:
- `/layers/<buildpack_name>/python`:
  - A layer containing the custom relocated Python installation (compiled in advanced and downloaded from S3)
- `/layers/<buildpack_name>/pip`:
  - A layer containing pip, as a user site-packages install (using `PYTHONUSERBASE` to set the location)
  - This layer can then be marked as `launch=false` to exclude it from the run image
- `/layers/<buildpack_name>/venv`:
  - A layer containing the application dependencies in a PEP-405 compliant venv
  - `PIP_PYTHON` points at this layer, so `pip install` will install into it rather than the global Python installation
- `/workspace` (the location of which is end-user customisable, and often changed to eg `/app`)
  - The layer containing the app source code

Each layer can then have its own cache invalidation logic. Any layers that are unchanged don't have to be rebuilt or even pushed to the remote. More at:
https://buildpacks.io/docs/for-buildpack-authors/concepts/
https://github.com/buildpacks/spec/blob/main/buildpack.md

Use of a venv for the app dependencies is needed due to:
1. A system install not allowing keeping the dependencies separate for cache invalidation
2. `--user` installs being problematic for relocated Python (xref https://github.com/astral-sh/uv/issues/2077#issuecomment-2312410060)
3. `PYTHONPATH` tricks being problematic due to stdlib shadowing (paths on `PYTHONPATH` take precedence over the stdlib, which can cause hard to debug issues from eg outdated backport packages in the app's transitive dependency tree)

As such, when we add support for uv in the future, we'll need a way to force the venv to be created outside the project directory and in its own layer. 

I'm mostly indifferent as to whether we have to create the venv ourselves or whether uv does that for us (so long as we can control the exact path uv creates; a generated venv path name derived from the project like Poetry does would be more hassle). If manually created, I also don't mind too much how we'd tell uv to use that venv - for pip we'll soon be using `PIP_PYTHON` (since pip doesn't check `VIRTUAL_ENV`), and for the upcoming Poetry support we instead set `VIRTUAL_ENV`.

**Comment by adiberk** [comment]
@zanieb The guide is pretty thorough and helpful! I was actually able to get it working using a watch config which I guess is fine though it seems inconsistent sometimes and may not work as well as gunicorns hot reload setup! 

I will say that I did try using the new setup without watch config and trying to mount .:/app but the app failed due to loosing my env it seems.

Regardless - really appreciate your help in this!

**Comment by zanieb** [comment]
@edmorley thank you for your thorough response. I take that feedback pretty seriously and it makes me think we'll need to expose _some_ way to allow custom virtual environment locations. The difficulty will be in encouraging users that would be better suited _not_ customizing the location on the happy path.

@adiberk I'm glad it helped! What's inconsistent about the watch setup? What happened with the mount? Feel free to open a targeted issue if something from the documentation didn't work out and we can chat over there.

**Comment by hynek** [comment]
For those who care about fast multi-stage builds, I've written down my own workflow here: https://hynek.me/articles/docker-uv/

tl;dr the lack of this feature in my context is awkward but easy to work around.

**Comment by tiangolo** [comment]
I came back just to say: thanks for writing this guide! https://docs.astral.sh/uv/guides/integration/docker/#configuring-watch-with-docker-compose :tada: 

I didn't know about Docker Compose's `watch`, that solves my problems even better than what I had (bind mounts). I learned something, thanks! :nerd_face: :rocket:

**Comment by zanieb** [comment]
Great to hear! Thank you!

**Comment by adiberk** [comment]
@zanieb I am 85% sure this is not a uv issue, but I figure why not double check.
I had to continue with my setup where I used an alternative working directory for creating the virtual env and then setting the virtual evn in the PATH in my docker file. However once I added a new system user that isn't root, I started hitting an issue. 
EX: `RUN useradd --create-home appuser... USER appuser`
Now when running my app
bash: /virtualenv/.venv/bin/uvicorn: /virtualenv/.venv/bin/python: bad interpreter: Permission denied

Even after setting the right permissions for this user to access the virtual env folder etc. 
```
# Prepare install
RUN mkdir -p /virtualenv && chown lev:lev /virtualenv
WORKDIR /virtualenv
```
This works absolutely fine though if I am using root user. Do you know if this is a UV thing or is this simply something I have to work on with my team?
